libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2259,18 +2258,23 @@ string YulUtilFunctions::readFromMemoryOrCalldata(Type const& _type, bool _fromC
 				("splitFunction", splitExternalFunctionIdFunction())
 				.render();
 
+
 		return Whiskers(R"(
-			function <functionName>(memPtr) -> value {
-				value := <load>(memPtr)
-				<?needsValidation>
+			function <functionName>(ptr) -> value {
+				<?fromCalldata>
+					value := calldataload(ptr)
 					<validate>(value)
-				</needsValidation>
+				<!fromCalldata>
+					value := <cleanup>(mload(ptr))
```

Comment:
Do we actually need to do cleanup after memory load now if we do it before operations and before memory store?

Reply:
The whole cleanup business as it currently stands relies on the optimizer to remove unnecessary cleanups anyways - so given that I'd say better to clean one time too many than not enough...

Actually I'm still looking at the cleanup before the memory store - it kind of looks like there could be a way to circumvent cleaning up there, but I haven't managed to trigger a dirty write yet, so maybe it's fine as is.

Reply:
I think cleanup before memory store is more important than cleanup after memory read because of things like `sha3` and external calls.

Reply:
Ok, I'll make double-sure there's cleanup there, resp. just add another explicit one to it.
It's a bit unfortunate that the IRVariable mechanism was kind of designed with the idea of things being clean in mind, so they can be assigned to each other all the time without problems, if they are of the same type... given that we change the cleanup policy, maybe we should rethink that...

Reply:
Actually cleanup is *always* done in ``writeToMemoryFunction``, so this is fine.
And it makes sense to generally not clean up IRVariables, but have the util functions *always* do it.